File: rag/domain/multilingual.py
"""The multilingual answering layer: Hindi, Tamil, and Gujarati.

The product serves several Supported Languages over one English Source of Truth.
This module is the seam that lets it: it detects the user's language, extracts the
query's intent into an English string with legal terms preserved (so retrieval and
reasoning always run over the single English corpus), and renders critical terms
back into the user's language for the answer.

The Bilingual Legal Glossary is the deterministic backbone here: the same curated
table both normalises an incoming (or code-mixed) query to English and constrains
the terminology in the output, so a term like bailable versus non-bailable cannot
flip meaning in translation. The layer is data-driven by language code - adding a
language is adding a script row and a glossary column, not new branches - and it
flags terms whose translation lacks an official source, the known soft spot for
lower-resource languages like Tamil and Gujarati. The live intent extractor
(:class:`rag.infrastructure.llm.LLMIntentExtractor`) implements the protocol
below with the glossary terms injected as hard constraints; the product has no
offline extractor (ADR 0010) - tests inject their own doubles.
"""
from __future__ import annotations

import json
import os
import re
from dataclasses import dataclass
from typing import Dict, List, Optional, Protocol, Sequence

from rag.domain.text import content_stems

ENGLISH = "en"
HINDI = "hi"
TAMIL = "ta"
GUJARATI = "gu"

_GLOSSARY_PATH = os.path.join(
    os.path.dirname(os.path.dirname(os.path.dirname(__file__))), "data", "glossary.json"
)

# Each non-English Supported Language writes in its own Unicode block, and that
# script is what marks a query as that language rather than English. Code-mixing
# (e.g. Hinglish) still trips the right script as long as any word is in it; the
# Latin-script tokens around it are preserved verbatim. Detection walks these in
# order, so adding a language is adding one row here and its glossary column.
_SCRIPTS: Sequence[tuple] = (
    (HINDI, re.compile(r"[ऀ-ॿ]")),  # Devanagari
    (TAMIL, re.compile(r"[஀-௿]")),  # Tamil
    (GUJARATI, re.compile(r"[઀-૿]")),  # Gujarati
)

# Any supported non-Latin script: a token carrying one is a foreign-script word.
# Used to drop function words the glossary does not carry once legal terms have
# been mapped to English, so retrieval always runs over English-only text.
_FOREIGN_RE = re.compile(
    "|".join(pattern.pattern for _, pattern in _SCRIPTS)
)


@dataclass(frozen=True)
class GlossaryEntry:
    """One critical legal term and its equivalents across Supported Languages.

    ``by_language`` maps a Supported Language code to the term in that language.
    ``unverified`` names the languages whose translation lacks an official
    central-act source and is therefore flagged for review.
    """

    en: str
    by_language: Dict[str, str]
    unverified: frozenset = frozenset()


def has_foreign_script(text: str) -> bool:
    """Whether any supported non-Latin script appears in ``text``.

    The live intent extractor uses this to skip the LLM call for a pure-English
    query; the glossary uses the same scripts to drop unmapped foreign tokens.
    """
    return bool(_FOREIGN_RE.search(text))


def detect_language(query: str) -> str:
    """Detect the user's language from the raw query.

    Returns the code of the first Supported Language whose script appears in the
    query (including code-mixed queries), otherwise ``"en"``. Detection drives the
    language of the answer.
    """
    for language, pattern in _SCRIPTS:
        if pattern.search(query):
            return language
    return ENGLISH


class BilingualGlossary:
    """The curated table of critical legal terms across Supported Languages.

    It serves two directions over the same rows: ``to_english`` rewrites a Hindi or
    code-mixed query into an English query for retrieval, and ``render`` renders an
    English term back into the user's language with the English kept inline in
    brackets so the critical term cannot be lost in translation.
    """

    def __init__(self, entries: Sequence[GlossaryEntry]):
        self._entries = list(entries)
        # Forward maps, one per Supported Language: English term -> target term.
        self._en_to: Dict[str, Dict[str, str]] = {}
        for entry in self._entries:
            for language, term in entry.by_language.items():
                self._en_to.setdefault(language, {})[entry.en] = term
        # One reverse map for normalisation: every foreign term -> its English
        # equivalent, longest foreign forms first so a phrase ("चल संपत्ति",
        # "அசையும் சொத்து") is matched before any of its component words.
        self._foreign_to_en = sorted(
            (
                (term, entry.en)
                for entry in self._entries
                for term in entry.by_language.values()
            ),
            key=lambda pair: len(pair[0]),
            reverse=True,
        )

    @classmethod
    def load(cls, path: str = _GLOSSARY_PATH) -> "BilingualGlossary":
        with open(path, "r", encoding="utf-8") as handle:
            raw = json.load(handle)
        rows = list(raw.get("terms", [])) + list(raw.get("domains", []))
        return cls([cls._entry_from_row(row) for row in rows])

    @staticmethod
    def _entry_from_row(row: Dict) -> GlossaryEntry:
        # Every key other than the English term and the bookkeeping ``unverified``
        # flag is a Supported Language column, so new languages are pure data.
        by_language = {
            key: value
            for key, value in row.items()
            if key not in ("en", "unverified")
        }
        return GlossaryEntry(
            en=row["en"],
            by_language=by_language,
            unverified=frozenset(row.get("unverified", [])),
        )

    def to_english(self, query: str) -> str:
        """Normalise a query to an English query, preserving legal terms.

        Each known foreign term (longest first) is replaced with its English
        equivalent; remaining foreign-script tokens (function words the glossary
        does not carry) are dropped, while Latin-script tokens - English words mixed
        into a code-mixed query - are kept verbatim. An English query passes through
        unchanged.
        """
        text = query
        for foreign, en in self._foreign_to_en:
            text = text.replace(foreign, f" {en} ")
        tokens = [t for t in text.split() if not _FOREIGN_RE.search(t)]
        return " ".join(tokens)

    def term_for(self, english_term: str, language: str) -> Optional[str]:
        """The glossary's equivalent of an English term in ``language``, if any."""
        return self._en_to.get(language, {}).get(english_term)

    def constraints_for(self, language: str) -> Dict[str, str]:
        """Foreign term -> authoritative English term for a Supported Language.

        These are the deterministic hard constraints injected into the LLM
        intent-extraction prompt so a normalised query lands on the glossary's
        English legal terms rather than whatever paraphrase the model would pick.
        The glossary stays the source of terminology; the LLM only rewrites around
        it.
        """
        return {
            entry.by_language[language]: entry.en
            for entry in self._entries
            if language in entry.by_language
        }

    def hindi_for(self, english_term: str) -> Optional[str]:
        """The glossary's Hindi equivalent of an English term, if it has one."""
        return self.term_for(english_term, HINDI)

    def render(self, english_term: str, language: str) -> str:
        """Render a critical term in ``language`` with the English inline in brackets.

        For a non-English language this yields ``"<term> (<english>)"`` so the
        critical legal term appears in the user's language while its authoritative
        English stays visible. Falls back to the English term when no equivalent
        exists or the language is English.
        """
        if language == ENGLISH:
            return english_term
        translated = self.term_for(english_term, language)
        return f"{translated} ({english_term})" if translated else english_term

    def unverified_terms(self, language: str) -> List[str]:
        """English terms whose ``language`` translation lacks an official source.

        Tamil and Gujarati lean on this: official central-act translations are
        often unavailable to verify a term, so those rows are flagged here rather
        than presented as authoritative.
        """
        return [e.en for e in self._entries if language in e.unverified]


@dataclass(frozen=True)
class NormalizedQuery:
    """A raw query after detection and intent extraction."""

    language: str
    english_query: str


class IntentExtractor(Protocol):
    """Detect the language and normalise a query to English for retrieval.

    This is the same live-only seam as the embedder and generator: production wires
    a live LLM behind it and there is no offline stand-in (ADR 0010); tests inject
    their own doubles. It returns a :class:`NormalizedQuery` whose English text
    drives every downstream step over the single English Source of Truth.
    """

    def normalize(self, query: str) -> NormalizedQuery: ...


# Ambiguous terms that, on their own, do not pin down a single Covered Domain.
# Such a query triggers a Confirmation Step rather than a guess. Each carries the
# clarifying question in every Supported Language.
_AMBIGUOUS: Dict[str, Dict[str, str]] = {
    "right": {
        ENGLISH: (
            "Did you mean your fundamental rights, your consumer rights, or your "
            "intellectual property rights? Please clarify so I can answer "
            "accurately."
        ),
        HINDI: (
            "क्या आपका मतलब आपके मौलिक अधिकार (fundamental rights), उपभोक्ता अधिकार "
            "(consumer rights), या बौद्धिक संपदा अधिकार (intellectual property rights) "
            "से है? कृपया स्पष्ट करें ताकि मैं सटीक उत्तर दे सकूँ।"
        ),
        TAMIL: (
            "நீங்கள் உங்கள் அடிப்படை உரிமைகள் (fundamental rights), நுகர்வோர் உரிமைகள் "
            "(consumer rights), அல்லது அறிவுசார் சொத்து உரிமைகள் (intellectual property "
            "rights) குறித்து கேட்கிறீர்களா? துல்லியமாக பதிலளிக்க தயவுசெய்து "
            "தெளிவுபடுத்தவும்."
        ),
        GUJARATI: (
            "શું તમારો અર્થ તમારા મૂળભૂત અધિકારો (fundamental rights), ગ્રાહક અધિકારો "
            "(consumer rights), અથવા બૌદ્ધિક સંપદા અધિકારો (intellectual property "
            "rights) છે? કૃપા કરીને સ્પષ્ટ કરો જેથી હું સચોટ જવાબ આપી શકું."
        ),
    },
}


def confirmation_for(english_query: str, language: str = ENGLISH) -> Optional[str]:
    """The Confirmation Step question for an ambiguous query, or ``None``.

    A query is ambiguous when its only content words are ambiguous terms - it
    carries no other legal content to disambiguate against. The clarifying check is
    returned in the user's language.
    """
    stems = set(content_stems(english_query))
    if not stems or not stems <= set(_AMBIGUOUS):
        return None
    for term in _AMBIGUOUS:
        if term in stems:
            texts = _AMBIGUOUS[term]
            return texts.get(language, texts[ENGLISH])
    return None

File: web/convex/lib/multilingual.ts
// The multilingual answering layer (Hindi, Tamil, Gujarati), ported from
// rag/domain/multilingual.py. Several Supported Languages are served over one
// English Source of Truth: the user's language is detected by script, the
// query's intent is extracted into English (legal terms preserved), and
// critical terms render back into the user's language.
//
// The Bilingual Legal Glossary is the deterministic backbone: the same
// curated table both normalises an incoming (or code-mixed) query to English
// and constrains the terminology of the output, so a term like bailable vs
// non-bailable cannot flip meaning in translation. The glossary data is
// bundled from data/glossary.json at build time.
import glossaryJson from "../../../data/glossary.json";
import { contentStems } from "./text";

export const ENGLISH = "en";
export const HINDI = "hi";
export const TAMIL = "ta";
export const GUJARATI = "gu";

export const SUPPORTED_LANGUAGES = [ENGLISH, HINDI, TAMIL, GUJARATI] as const;

// Each non-English Supported Language writes in its own Unicode block, and
// that script is what marks a query as that language rather than English.
// Code-mixing (e.g. Hinglish) still trips the right script as long as any
// word is in it. Detection walks these in order, so adding a language is
// adding one row here and its glossary column.
const SCRIPTS: Array<[string, RegExp]> = [
  [HINDI, /[ऀ-ॿ]/], // Devanagari
  [TAMIL, /[஀-௿]/], // Tamil
  [GUJARATI, /[઀-૿]/], // Gujarati
];

// Any supported non-Latin script: a token carrying one is a foreign-script
// word. Used to drop function words the glossary does not carry once legal
// terms have been mapped to English.
const FOREIGN_RE = new RegExp(
  SCRIPTS.map(([, pattern]) => pattern.source).join("|"),
);

// Whether any supported non-Latin script appears in `text`. The live intent
// extractor uses this to skip the LLM call for a pure-English query.
export function hasForeignScript(text: string): boolean {
  return FOREIGN_RE.test(text);
}

// Detect the user's language from the raw query: the first Supported
// Language whose script appears (including code-mixed queries), else "en".
export function detectLanguage(query: string): string {
  for (const [language, pattern] of SCRIPTS) {
    if (pattern.test(query)) return language;
  }
  return ENGLISH;
}

// One critical legal term and its equivalents across Supported Languages.
// `unverified` names the languages whose translation lacks an official
// central-act source and is therefore flagged for review.
export type GlossaryEntry = {
  en: string;
  byLanguage: Record<string, string>;
  unverified: Set<string>;
};

type GlossaryRow = { en: string; unverified?: string[] } & Record<
  string,
  unknown
>;

function entryFromRow(row: GlossaryRow): GlossaryEntry {
  // Every key other than the English term and the bookkeeping `unverified`
  // flag is a Supported Language column, so new languages are pure data.
  const byLanguage: Record<string, string> = {};
  for (const [key, value] of Object.entries(row)) {
    if (key !== "en" && key !== "unverified" && typeof value === "string") {
      byLanguage[key] = value;
    }
  }
  return {
    en: row.en,
    byLanguage,
    unverified: new Set(row.unverified ?? []),
  };
}

// The curated table of critical legal terms across Supported Languages. It
// serves two directions over the same rows: toEnglish rewrites a Hindi or
// code-mixed query into an English query for retrieval, and render renders an
// English term back into the user's language with the English kept inline.
export class BilingualGlossary {
  private entries: GlossaryEntry[];
  private enTo: Map<string, Map<string, string>>;
  private foreignToEn: Array<[string, string]>;

  constructor(entries: GlossaryEntry[]) {
    this.entries = entries;
    // Forward maps, one per Supported Language: English term -> target term.
    this.enTo = new Map();
    for (const entry of entries) {
      for (const [language, term] of Object.entries(entry.byLanguage)) {
        if (!this.enTo.has(language)) this.enTo.set(language, new Map());
        this.enTo.get(language)!.set(entry.en, term);
      }
    }
    // One reverse map for normalisation: every foreign term -> its English
    // equivalent, longest foreign forms first so a phrase is matched before
    // any of its component words.
    this.foreignToEn = entries
      .flatMap((entry) =>
        Object.values(entry.byLanguage).map(
          (term) => [term, entry.en] as [string, string],
        ),
      )
      .sort((a, b) => b[0].length - a[0].length);
  }

  static load(): BilingualGlossary {
    const raw = glossaryJson as { terms?: GlossaryRow[]; domains?: GlossaryRow[] };
    const rows = [...(raw.terms ?? []), ...(raw.domains ?? [])];
    return new BilingualGlossary(rows.map(entryFromRow));
  }

  // Normalise a query to an English query, preserving legal terms. Each known
  // foreign term (longest first) is replaced with its English equivalent;
  // remaining foreign-script tokens are dropped, while Latin-script tokens -
  // English words mixed into a code-mixed query - are kept verbatim.
  toEnglish(query: string): string {
    let text = query;
    for (const [foreign, en] of this.foreignToEn) {
      text = text.split(foreign).join(` ${en} `);
    }
    return text
      .split(/\s+/)
      .filter((t) => t && !FOREIGN_RE.test(t))
      .join(" ");
  }

  // The glossary's equivalent of an English term in `language`, if any.
  termFor(englishTerm: string, language: string): string | null {
    return this.enTo.get(language)?.get(englishTerm) ?? null;
  }

  // Foreign term -> authoritative English term for a Supported Language: the
  // deterministic hard constraints injected into the LLM intent-extraction
  // prompt, so the curated glossary - not the model - fixes the terminology.
  constraintsFor(language: string): Record<string, string> {
    const constraints: Record<string, string> = {};
    for (const entry of this.entries) {
      const term = entry.byLanguage[language];
      if (term !== undefined) constraints[term] = entry.en;
    }
    return constraints;
  }

  // Render a critical term in `language` with the English inline in brackets,
  // so the critical legal term appears in the user's language while its
  // authoritative English stays visible.
  render(englishTerm: string, language: string): string {
    if (language === ENGLISH) return englishTerm;
    const translated = this.termFor(englishTerm, language);
    return translated ? `${translated} (${englishTerm})` : englishTerm;
  }

  // English terms whose `language` translation lacks an official source.
  unverifiedTerms(language: string): string[] {
    return this.entries
      .filter((e) => e.unverified.has(language))
      .map((e) => e.en);
  }
}

// A raw query after detection and intent extraction.
export type NormalizedQuery = {
  language: string;
  englishQuery: string;
};

// Detect the language and normalise a query to English for retrieval. The
// live implementation (convex/llm.ts) wires an LLM behind this; tests inject
// deterministic doubles. Same live-only seam as generation (ADR 0010).
export type IntentExtractor = (query: string) => Promise<NormalizedQuery>;

// Ambiguous terms that, on their own, do not pin down a single Covered
// Domain. Such a query triggers a Confirmation Step rather than a guess.
const AMBIGUOUS: Record<string, Record<string, string>> = {
  right: {
    [ENGLISH]:
      "Did you mean your fundamental rights, your consumer rights, or your " +
      "intellectual property rights? Please clarify so I can answer " +
      "accurately.",
    [HINDI]:
      "क्या आपका मतलब आपके मौलिक अधिकार (fundamental rights), उपभोक्ता अधिकार " +
      "(consumer rights), या बौद्धिक संपदा अधिकार (intellectual property rights) " +
      "से है? कृपया स्पष्ट करें ताकि मैं सटीक उत्तर दे सकूँ।",
    [TAMIL]:
      "நீங்கள் உங்கள் அடிப்படை உரிமைகள் (fundamental rights), நுகர்வோர் உரிமைகள் " +
      "(consumer rights), அல்லது அறிவுசார் சொத்து உரிமைகள் (intellectual property " +
      "rights) குறித்து கேட்கிறீர்களா? துல்லியமாக பதிலளிக்க தயவுசெய்து " +
      "தெளிவுபடுத்தவும்.",
    [GUJARATI]:
      "શું તમારો અર્થ તમારા મૂળભૂત અધિકારો (fundamental rights), ગ્રાહક અધિકારો " +
      "(consumer rights), અથવા બૌદ્ધિક સંપદા અધિકારો (intellectual property " +
      "rights) છે? કૃપા કરીને સ્પષ્ટ કરો જેથી હું સચોટ જવાબ આપી શકું.",
  },
};

// The Confirmation Step question for an ambiguous query, or null. A query is
// ambiguous when its only content words are ambiguous terms - it carries no
// other legal content to disambiguate against.
export function confirmationFor(
  englishQuery: string,
  language: string = ENGLISH,
): string | null {
  const stems = new Set(contentStems(englishQuery));
  if (stems.size === 0) return null;
  const ambiguousTerms = new Set(Object.keys(AMBIGUOUS));
  if (![...stems].every((s) => ambiguousTerms.has(s))) return null;
  for (const term of Object.keys(AMBIGUOUS)) {
    if (stems.has(term)) {
      const texts = AMBIGUOUS[term];
      return texts[language] ?? texts[ENGLISH];
    }
  }
  return null;
}
